packages/suite-desktop/src/Main.tsx

```diff
@@ -71,15 +71,20 @@ export const init = async (root: HTMLElement) => {
     const preloadAction = await preloadStore();
     const store = initStore(preloadAction);
 
-    await desktopApi.handshake();
+    const handshakeResponse = await desktopApi.handshake();
+    const { torSettings } = handshakeResponse;
 
-    // render more complex interactive loader with theme provider without redux
-    render(<ModulesLoadingScreen />, root);
+    render(<TorLoadingScreen torSettings={torSettings} />, root);
+
+    const loadTorModule = await desktopApi.loadTor(null);
+    if (!loadTorModule.success) {
+        await new Promise(resolve => {
+            render(<TorLoadingScreen torSettings={torSettings} callback={resolve} />, root);
```

Comment:
I would change property name `callback` to something like `skipTorLoading` (is it what it does right?)

Reply:
It is actually to make it wait until the "Tor initialization" is completed. 
Basically that will run if Tor loading fails and then we need to wait for user input (user has to decide whether to try again or disable) so since it is running in a function and not in a component we need a way to wait.

I know it is a bit weird but it is the solution to being able to wait for user input in a function that is not a react component.

Reply:
Maybe it looks a bit less controversial like this? But it's exactly the same.
```
if (!loadTorModule.success) {
    const { resolve, promise } = createDeferred();
    render(<TorLoadingScreen torSettings={torSettings} callback={resolve} />, root);
    await promise;
}
```

Reply:
@marekrjpolak I agree, it is less weird like you suggested  5b1f42ebed895deebfc245dcde83c7bc1dc26aa0

Reply:
> And now I think we could use `createDeferred` even for rendering `<TorLoadingScreen />` only once (which was another thing I didn't like much). Something like this:
> 
> ```
> const { resolve, promise } = createDeferred();
> render(<TorLoadingScreen torSettings={torSettings} callback={resolve} />, root);
> const loadTorModule = await desktopApi.loadTor();
> if (!loadTorModule.success) {
>      await promise;
> }
> ```

Isn't it more complicated to understand what is really happening?

Reply:
Well, it could be. But to be honest, I was also a bit confused why we render the same component twice and why we `setError` when `callback` is passed. I'm thinking how to make the potentially repeating process better understandable without the need of rerendering. :thinking:

Reply:
I agree with that, but we need to know inside `<TorLoadingScreen>` that `loadTorModule` was not successful and yes we could listen to `tor/bootstrap` errors but, what about the errors that are happening because there was not connection
 to ControlPort for example? 

Maybe it could be possible if we call `desktopApi.loadTor();` inside the `TorLoadingScreen` component or if we pass the promise as a prop and then check inside if it has failed, but we are doing weird things again.

Reply:
Some input from @szymonlesisz:
- prefers the `await new Promise(resolve => render(<TorLoadingScreen callback={resolve} />, root))` variant, so we have a stalemate situation and you decide
- suggests resolving the double-render by calling `desktopApi.loadTor()` from inside (exactly as you proposed) but we'd like to pass the function as a prop (for better readability), therefore probably something like `<TorLoadingScreen loadTor={() => desktopApi.loadTor()} resolve={resolve} />`

Reply:
i was just about to write that :) thanks @marekrjpolak :)
its just my opinion, but written like this its obvious what exactly it does. It is waiting for promise which is resolved inside rendered component. you dont need to walk thru code to see how `createDeferred` works, or why is there double renderer etc.

`TorLoadingScreen` will do everything it needs to be done, will call desktopApi, will use local state etc. and after job is done it call `resolve` and will be unmonuted

Reply:
Ok, so I think it should be :ok_hand:  now.
I took the `await new Promise(...` approach since it is true that makes it more simple to read for someone not familiar with `createDeferred` util. But I do not have hard opinions on that.

But I think that initializing the Tor module before rendering the component to know if it should run or not it is better because it allows us just not to render the TorLoadingScreen when there is no need. Other wise we should have to render a Loading component inside and wait for response.